### ISSUE TITLE
Issue with BasemapGallery and source autocast

### ISSUE THREAD

**Issue by jaouenk** [Esri/jsapi-resources#100]
Hi, with the lastest API version 4.5.0, i have a issue with the autocast for the source parameter of the BasemaGallery component and TypeScript.

Here a sample :
 ```
let basemapGallery = new BasemapGallery({
      container: this.container.nativeElement,
      view: this.mapView
    });
const BasemapCollection = Collection.ofType(Basemap);
let collection: Collection<Basemap> = new BasemapCollection();
// fill the collection with some basemaps
basemapGallery.source = collection;
```
It works but I have this TS error message : 

> TS2322: Type 'Collection<Basemap>' is not assignable to type 'LocalBasemapsSource | PortalBasemapsSource'.
>   Type 'Collection<Basemap>' is not assignable to type 'PortalBasemapsSource'.
>     Property 'basemaps' is missing in type 'Collection<Basemap>'.

In the documentation, it's indicate that we can set this property with an array or Collection of Basemaps
[https://developers.arcgis.com/javascript/latest/api-reference/esri-widgets-BasemapGallery.html#source](url)

Is there any feedback about the autocast and TypeScript ?

**Comment by jaouenk** [comment]
This line bypass the problem : 
`basemapGallery.source = (collection as any) as LocalBasemapsSource;`

**Comment by jkieboom** [comment]
It is unfortunately not possible for us to support auto-cast in typescript 100%. In particular, we cannot support it typings wise on property setters because TS does not allow us to type the setter different from the getter. This is quite unfortunate, but something that we can't do anything about. You can either:

1. Cast to `any` as you've seen (we don't recommend this)
1. Use `gallery.set("source", collection)`. The `set` function should allow you to auto-cast, we may provide strict typings for `.set` in the future.
1. Avoid using auto-cast in TS when setting properties (i.e. just use `gallery.source = new LocalBasemapsSource({ basemaps: collection })`)

3) Is currently what we would currently recommend in TS

**Comment by jaouenk** [comment]
Thanks for your answer. 

Your third point doesn't work either. We have this error inside the console : 
TS2351: Cannot use 'new' with an expression whose type lacks a call or construct signature.
**Sample :** 
`basemapGallery.source = new LocalBasemapsSource({ basemaps : collection});`

We will use the set function. It works and it seems to be the best solution. Thank you

**Comment by jkieboom** [comment]
Regard point 3) this is surprising to me, it seems like these are not documented as actual classes and I don't know the reason for that. /cc @bsvensson ?

**Comment by ekenes** [comment]
@jkieboom it was unintentional. We'll fix it.

**Comment by dasa** [comment]
> these are not documented as actual classes

Fixed in 4.6
